---
title: "NHP Amplicons Sequencing Analysis"
author: "Meng Wang"
date: "`r format(Sys.time(), '%B %d, %Y')`"
execute:
        cache: true
format:
        html:
            code-fold: true
            code-overflow: "wrap"
            embed-resources: true
            tidy: true
            code-link: false
editor_options:
  chunk_output_type: console
---



```{r}
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(gt))
suppressPackageStartupMessages(library(patchwork))
```

# Alignment Algorithm
- Finding sequences that matches the (truncated) upstream-{NNK}-downstream pattern.
    - for both forward and reverse strand.
- Extract the 7X NNK repeats.
- Only retain NNK repeats starting at 40bp (forward strand) and 115bp (reverse strand).
- Only retain NNK repeats that are in agreement between R1 and R2.
- Only retain NNK repeats that have fread score > 30.

# QC

## AAC9
- Upstream: GAGTGCCCAA
- Downstream: GCACAGGCGC

```{r}
#| eval: FALSE
#| include: TRUE
# processing and filtering
# Input
r1  <- fread('NHP-input-6-9-10-11152025_R1_001.txt') %>%
    mutate(ReadName = stringr::str_split_i(ReadHeader, pattern=" ", 1)) %>%
    select(-ReadHeader)

r2 <- fread('NHP-input-6-9-10-11152025_R2_001.txt') %>%
    mutate(ReadName = stringr::str_split_i(ReadHeader, pattern=" ", 1)) %>%
    select(-ReadHeader)

df <- merge(r1, r2, by = "ReadName", all = TRUE)


# 97.05% of reads agree on alignment

df1 <- subset(df, DNASequence.x==DNASequence.y) %>%
       subset((StartPosition.x == 40 & Orientation.x == 'forward') | (StartPosition.x == 115 | Orientation.x == 'reverse')) %>%
       subset((StartPosition.y == 40 & Orientation.y == 'forward') | (StartPosition.y == 115 | Orientation.y == 'reverse')) %>%
       subset(AvgQuality.x > 30 & AvgQuality.y > 30) %>%
       subset(Orientation.x != Orientation.y) %>%
       select(-DNASequence.y) %>%
       select(-PeptideSequence.y) %>%
       select(-Orientation.y) %>%
       select(-AvgQuality.y) %>%
       select(-StartPosition.y)
    #    subset(ValidNNK.y & ValidNNK.x)

fwrite(df1, 'input_clean.txt', sep="\t")

## Treated
r1  <- fread('NHP-SC-6-9-10-11142025_R1_001.txt') %>%
    mutate(ReadName = stringr::str_split_i(ReadHeader, pattern=" ", 1)) %>%
    select(-ReadHeader)

r2 <- fread('NHP-SC-6-9-10-11142025_R2_001.txt') %>%
    mutate(ReadName = stringr::str_split_i(ReadHeader, pattern=" ", 1)) %>%
    select(-ReadHeader)

df <- merge(r1, r2, by = "ReadName", all = TRUE)


# 97.05% of reads agree on alignment

df1 <- subset(df, DNASequence.x==DNASequence.y) %>%
       subset((StartPosition.x == 40 & Orientation.x == 'forward') | (StartPosition.x == 115 | Orientation.x == 'reverse')) %>%
       subset((StartPosition.y == 40 & Orientation.y == 'forward') | (StartPosition.y == 115 | Orientation.y == 'reverse')) %>%
       subset(AvgQuality.x > 30 & AvgQuality.y > 30) %>%
       subset(Orientation.x != Orientation.y) %>%
       select(-DNASequence.y) %>%
       select(-PeptideSequence.y) %>%
       select(-Orientation.y) %>%
       select(-AvgQuality.y) %>%
       select(-StartPosition.y)
    #    subset(ValidNNK.y & ValidNNK.x)

fwrite(df1, 'SC_clean.txt', sep="\t")
```


```{r}
treated <- fread('SC_clean.txt', sep="\t")
input <- fread('input_clean.txt', sep="\t")

qc <- read.table('/igm/home/mxw010/NHP/analysis/QC.txt', header=T) %>%
  mutate(Final_NNK = rep(c(sum(treated$ValidNNK.x==TRUE), sum(input$ValidNNK.x==TRUE)), each=2))

qc[,-c(3, 7)] %>%
  rename(NNK=Matching_Flanking_Sequences) %>%
  rename(Reads=Total_Reads) %>%
  gt() %>%
  fmt_number(columns = 2:6, decimals = 0)
```

## Peptide Counts
- peptides with stop codons removed.
- invalid NNK sequences not removed.
```{r}
sc_count <- length(unique(treated$PeptideSequence.x[-grep("\\*", treated$PeptideSequence.x)]))
input_count <- length(unique(input$PeptideSequence.x[-grep("\\*", treated$PeptideSequence.x)]))
tibble(sample = c('SC', 'Input'), counts=c(sc_count, input_count), .rows = 1) %>%
  gt(rowname_col="sample") %>%
  cols_label(counts = "Unique Peptide Counts") %>%
  fmt_number(decimals=0)

```
## Distribution of Peptides (<200)
The lower ends mostly include noise and should be further filtered before analysis. Looking at distribution of peptides counts, repeats with less than 50 observations are removed.
```{r}
p1 <- subset(treated, ValidNNK.x & ValidNNK.y) %>% 
        group_by(PeptideSequence.x) %>%
        summarize(count=n(), .groups='drop') %>%
        subset(count < 200) %>%
        ggplot(aes(x=count)) +
        geom_histogram(bins=100) + 
        xlab('NNK Repeats Counts') +
        ylab('Frequency') +
        ggtitle("Treated, valid NNK")

p2 <- subset(treated, !(ValidNNK.x & ValidNNK.y)) %>% 
        group_by(PeptideSequence.x) %>%
        summarize(count=n(), .groups='drop') %>%
        subset(count < 200) %>%
        ggplot(aes(x=count)) +
        geom_histogram(bins=100) +
        xlab('NNK Repeats Counts') +
        ggtitle("Treated, invalid NNK")

p3 <- subset(input, ValidNNK.x & ValidNNK.y) %>% 
        group_by(PeptideSequence.x) %>%
        summarize(count=n(), .groups='drop') %>%
        subset(count < 200) %>%
        ggplot(aes(x=count)) +
        geom_histogram(bins=100) + 
        xlab('NNK Repeats Counts') +
        ylab('Frequency') +
        ggtitle("Input, valid NNK")

p4 <- subset(input, !(ValidNNK.x & ValidNNK.y)) %>% 
        group_by(PeptideSequence.x) %>%
        summarize(count=n(), .groups='drop') %>%
        subset(count < 200) %>%
        ggplot(aes(x=count)) +
        geom_histogram(bins=100) +
        xlab('NNK Repeats Counts') +
        ggtitle("Input, invalid NNK")
  
p1 + p2 + p3 + p4 + plot_layout(ncol=2)
```

## Distribution of NNK Counts (>50)
For Treated, invalid, only one peptides are observed more than 100 times (RGDLGLS).

For valid NNK barcodes, the distribution of number of observations follow the exponential decay law. 

The figures based on DNA sequences are very similar and included at the end of the page.
```{r}
p1 <- subset(treated, ValidNNK.x & ValidNNK.y) %>% 
        group_by(PeptideSequence.x) %>%
        summarize(count=n(), .groups='drop') %>%
        subset(count > 50) %>%
        ggplot(aes(x=count)) +
        geom_histogram(bins=100) + 
        xlab('NNK Repeats Counts') +
        ylab('Frequency') +
        ggtitle("Treated, valid NNK")

p2 <- subset(treated, !(ValidNNK.x & ValidNNK.y)) %>% 
        group_by(PeptideSequence.x) %>%
        summarize(count=n(), .groups='drop') %>%
        subset(count > 50) %>%
        ggplot(aes(x=count)) +
        geom_histogram(bins=100) +
        xlab('NNK Repeats Counts') +
        ggtitle("Treated, invalid NNK")

p3 <- subset(input, ValidNNK.x & ValidNNK.y) %>% 
        group_by(PeptideSequence.x) %>%
        summarize(count=n(), .groups='drop') %>%
        subset(count > 50) %>%
        ggplot(aes(x=count)) +
        geom_histogram(bins=100) + 
        xlab('NNK Repeats Counts') +
        ylab('Frequency') +
        ggtitle("Input, valid NNK")

p4 <- subset(input, !(ValidNNK.x & ValidNNK.y)) %>% 
        group_by(PeptideSequence.x) %>%
        summarize(count=n(), .groups='drop') %>%
        subset(count > 50) %>%
        ggplot(aes(x=count)) +
        geom_histogram(bins=100) +
        xlab('NNK Repeats Counts') +
        ggtitle("Input, invalid NNK")
  
p1 + p2 + p3 + p4 + plot_layout(ncol=2)
```


## Top Peptides in NHP-SC
```{r}
treated <- treated[-grep("\\*", treated$PeptideSequence.x),]
treated %>% group_by(ValidNNK.x, PeptideSequence.x) %>%
  summarize(Counts = n(), .groups='drop') %>%
  arrange(desc(Counts)) %>%
  head(20) %>%
  rename(Peptide = PeptideSequence.x) %>%
  rename(ValidNNK = ValidNNK.x) %>%
  relocate(ValidNNK, .after=last_col()) %>%
  gt() %>%
  fmt_number(columns=Counts, decimals=0)

```

## Top 20 Enriched Peptides
- Removed any entry that contains stop codon
- Only retains peptides with counts > 50
- Enrichment is calculated as normalized counts (per sample) per million reads.
- If the peptide is not observed in input, then it is replaced with a count of 1 to avoid divison by 0.
- Result is saved in the enrichment column.
```{r}
# low_counts_peptides <- treated %>%
#     group_by(PeptideSequence.x) %>%
#     summarize(Counts = n()) %>%
#     subset(Counts > 50) %>%
#     pull(PeptideSequence.x)

treated_sub <- treated %>%
                rename(ValidNNK = ValidNNK.x) %>%
                  rename(PeptideSequence = PeptideSequence.x) %>%
                  group_by(ValidNNK, PeptideSequence) %>%
                  summarize(group_count = n(),
                             .groups='drop') %>%
                  subset(group_count > 50) %>%
                  mutate(normalized_count = group_count / sum(group_count ) * 10^6) %>%
                  mutate(ID = paste0(PeptideSequence, ValidNNK))
                  

# low_counts_peptides <- input %>%
#     group_by(PeptideSequence.x) %>%
#     summarize(Counts = n()) %>%
#     subset(Counts > 50) %>%
#     pull(PeptideSequence.x)

input_sub <- input %>%
                  rename(ValidNNK = ValidNNK.x) %>%
                  rename(PeptideSequence = PeptideSequence.x) %>%
                  group_by(ValidNNK, PeptideSequence) %>%
                  summarize(group_count = n(),
                             .groups='drop') %>%
                  mutate(normalized_count = group_count / sum(group_count ) * 10^6) %>%
                  mutate(ID = paste0(PeptideSequence, ValidNNK))


 df <- merge(treated_sub, input_sub, by = 'ID', all.x = TRUE, all.y = FALSE) %>%
      replace_na(list(group_count.y = 1, normalized_count.y = 1/sum(input_sub$group_count)*10^6)) %>%
      mutate(Enrichment = normalized_count.x/normalized_count.y) %>%
      mutate(Input=ifelse(is.na(ValidNNK.y), 'NA', group_count.y)) %>%
      select(PeptideSequence.x, group_count.x, Input, Enrichment, ValidNNK.x) %>%
      rename(PeptideSequence = PeptideSequence.x) %>%
   rename(SC = group_count.x) %>%
   rename(ValidNNK = ValidNNK.x) %>%
   arrange(desc(Enrichment)) 
 
df %>% 
  head(20) %>%
  gt() %>%
  fmt_number(columns=2:3, decimals=0)

# fwrite(df, 'AAV9_enrichment.csv', quote=FALSE, sep=",")
   
```


## Distribution of NNK Repeats DNA Sequences (<200)
```{r}
p1 <- subset(treated, ValidNNK.x & ValidNNK.y) %>% 
        group_by(DNASequence.x) %>%
        summarize(count=n(), .groups='drop') %>%
        subset(count < 200) %>%
        ggplot(aes(x=count)) +
        geom_histogram(bins=100) + 
        xlab('NNK Repeats Counts') +
        ylab('Frequency') +
        ggtitle("Treated, valid NNK")

p2 <- subset(treated, !(ValidNNK.x & ValidNNK.y)) %>% 
        group_by(DNASequence.x) %>%
        summarize(count=n(), .groups='drop') %>%
        subset(count < 200) %>%
        ggplot(aes(x=count)) +
        geom_histogram(bins=100) +
        xlab('NNK Repeats Counts') +
        ggtitle("Treated, invalid NNK")

p3 <- subset(input, ValidNNK.x & ValidNNK.y) %>% 
        group_by(DNASequence.x) %>%
        summarize(count=n(), .groups='drop') %>%
        subset(count < 200) %>%
        ggplot(aes(x=count)) +
        geom_histogram(bins=100) + 
        xlab('NNK Repeats Counts') +
        ylab('Frequency') +
        ggtitle("Input, valid NNK")

p4 <- subset(input, !(ValidNNK.x & ValidNNK.y)) %>% 
        group_by(DNASequence.x) %>%
        summarize(count=n(), .groups='drop') %>%
        subset(count < 200) %>%
        ggplot(aes(x=count)) +
        geom_histogram(bins=100) +
        xlab('NNK Repeats Counts') +
        ggtitle("Input, invalid NNK")
  
p1 + p2 + p3 + p4 + plot_layout(ncol=2)
```


## Distribution of NNK Repeats DNA Sequences (>50)

```{r}
p1 <- subset(treated, ValidNNK.x & ValidNNK.y) %>% 
        group_by(DNASequence.x) %>%
        summarize(count=n(), .groups='drop') %>%
        subset(count > 50) %>%
        ggplot(aes(x=count)) +
        geom_histogram(bins=100) + 
        xlab('NNK Repeats Counts') +
        ylab('Frequency') +
        ggtitle("Treated, valid NNK")

p2 <- subset(treated, !(ValidNNK.x & ValidNNK.y)) %>% 
        group_by(DNASequence.x) %>%
        summarize(count=n(), .groups='drop') %>%
        subset(count > 50) %>%
        ggplot(aes(x=count)) +
        geom_histogram(bins=100) +
        xlab('NNK Repeats Counts') +
        ggtitle("Treated, invalid NNK")

p3 <- subset(input, ValidNNK.x & ValidNNK.y) %>% 
        group_by(DNASequence.x) %>%
        summarize(count=n(), .groups='drop') %>%
        subset(count > 50) %>%
        ggplot(aes(x=count)) +
        geom_histogram(bins=100) + 
        xlab('NNK Repeats Counts') +
        ylab('Frequency') +
        ggtitle("Input, valid NNK")

p4 <- subset(input, !(ValidNNK.x & ValidNNK.y)) %>% 
        group_by(DNASequence.x) %>%
        summarize(count=n(), .groups='drop') %>%
        subset(count > 50) %>%
        ggplot(aes(x=count)) +
        geom_histogram(bins=100) +
        xlab('NNK Repeats Counts') +
        ggtitle("Input, invalid NNK")
  
p1 + p2 + p3 + p4 + plot_layout(ncol=2)
```

